joern-server/src/main/scala/io/shiftleft/joern/server/JoernController.scala

```diff
@@ -0,0 +1,158 @@
+  protected val applicationDescription = "Joern-Server REST API"
+  protected implicit def executor: ExecutionContext = system.dispatcher
+
+  private var cpg: Option[Cpg] = None
+  private var queryResult: Option[String] = None
+
+  before() {
+    contentType = formats("json")
+  }
+
+  private val createBuilder =
+    (apiOperation[Unit]("create")
+      summary "Create a code property graph"
+      tags ""
+      parameter queryParam[List[String]]("filenames").description("File/Directory names"))
+
+  post("/create", operation(createBuilder)) {
+    handleOrReportAsInvalid[CreateCpgRequest] { v =>
+      val filenames = v.filenames
+      if (filenames.isEmpty) {
+        halt(400, "`filenames` not given or invalid")
+      }
+      if (filenames.count(File(_).exists) != filenames.size) {
+        halt(400, "Not all specified files exist")
+      }
+      createCpg(filenames)
+      response.setHeader("Location", s"/status")
+      Accepted()
+    }
+
+    def createCpg(filenames: List[String]): AsyncResult =
+      new AsyncResult {
+        val is = Future {
+          val cpgFilename = "/tmp/cpg.bin.zip"
+          logger.info(s"Attempting to create CPG for: ${filenames.mkString(",")}")
+          JoernParse.parse(filenames.toArray, cpgFilename)
+          cpg = Some(CpgLoader.load(cpgFilename))
+          logger.info("CPG is ready")
+        }
+      }
+  }
+
+  private val statusBuilder = (apiOperation[String]("status")
+    summary "Status of the CPG"
+    tags "")
+
+  get("/status", operation(statusBuilder)) {
+    Ok(StatusResponse(isCpgLoaded = cpg.isDefined))
+  }
+
+  private val queryBuilder = (apiOperation[String]("query")
+    summary "query"
+    tags "")
+
+  post("/query", operation(queryBuilder)) {
+    handleOrReportAsInvalid[QueryRequest] { v =>
+      if (!cpg.isDefined) {
+        BadRequest("CPG is not loaded")
+      } else {
+        val query = v.query
+        logger.info(s"received query: $query")
+        runQuery(query)
+        Accepted()
+      }
+    }
+
+    def runQuery(query: String): AsyncResult =
+      new AsyncResult {
+        val is = Future {
+          Try({
+            logger.info("Running query")
+            import javax.script.ScriptEngineManager
+            val e = new ScriptEngineManager().getEngineByName("scala")
+            e.put("aCpg", cpg.get)
+            e.eval(s"""
+                import io.shiftleft.codepropertygraph.Cpg
+              | val cpg = aCpg.asInstanceOf[io.shiftleft.codepropertygraph.Cpg]
+              | $query
+            """.stripMargin).toString
+          }) match {
+            case Success(v)         => queryResult = Some(v)
+            case Failure(exception) => println(exception)
+          }
+
+        }
+      }
+  }
+
+  val queryResultBuilder = (apiOperation[String]("queryresult")
+    summary "query result"
+    tags "")
+
+  get("/queryresult", operation(queryResultBuilder)) {
+    if (queryResult.isDefined) {
+      val result = queryResult.get
```

Comment:
Content-type was json, but you are right, it was possible that this would just return a non-json string. I've fixed this. We now return a JSON object of the form `{"response" : "string response", "isQueryCompleted" : true/false}`. The string response may still not be json - it is a string representation of whatever the query yields.